pennylane/resources/factorization.py

```diff
@@ -0,0 +1,141 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for two-electron tensor factorization.
+"""
+from pennylane import numpy as np
+
+
+def factorize(two, tol):
+    r"""Return the double-factorized form of a two-electron tensor.
+
+    The second quantized electronic Hamiltonian is constructed in terms of fermionic creation,
+    :math:`a^{\dagger}` , and annihilation, :math:`a`, operators as
+    [`arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`_]
+
+    .. math::
+
+        H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} h_{pq} a_{p,\alpha}^{\dagger}
+        a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
+        h_{pqrs} a_{p, \alpha}^{\dagger} a_{q, \beta}^{\dagger} a_{r, \beta} a_{s, \alpha},
+
+    where :math:`h_{pq}` and :math:`h_{pqrs}` are the one- and two-electron integrals computed as
+
+    .. math::
+
+        h_{pq} = \int \phi_p(r)^* \left ( -\frac{\nabla_r^2}{2} - \sum_i \frac{Z_i}{|r-R_i|} \right)
+        \phi_q(r) dr,
+
+    and
+
+    .. math::
+
+        h_{pqrs} = \int \frac{\phi_p(r_1)^* \phi_q(r_2)^* \phi_r(r_2) \phi_s(r_1)}{|r_1 - r_2|}
+        dr_1 dr_2.
+
+    The two-electron integrals can be rearranged in the so-called chemist notation which gives
+
+    .. math::
+
+        V_{pqrs} = \int \frac{\phi_p(r_1)^* \phi_q(r_1)^* \phi_r(r_2) \phi_s(r_2)}{|r_1 - r_2|}
+        dr_1 dr_2,
+
+    and the molecular Hamiltonian can be rewritten as
+
+    .. math::
+
+        H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} T_{pq} a_{p,\alpha}^{\dagger}
+        a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
+        V_{pqrs} a_{p, \alpha}^{\dagger} a_{q, \alpha} a_{r, \beta}^{\dagger} a_{s, \beta}.
+
+    with
+
+    .. math::
+
+        T_{pq} = h_{pq} - \frac{1}{2} \sum_s h_{pssq}.
+
+
+    This notation allows a low-rank factorization of the two-electron integral. The objective of
+    the factorization is to find a set of symmetric matrices, :math:`L`, such that
+
+    .. math::
+
+           V_{ijkl} = \sum_r^R L_{ij}^{(r)} L_{kl}^{(r) T},
+
+    with the rank :math:`R \leq n^2`. The matrices :math:`L` are diagonalized and for each matrix
+    the eigenvalues that are smaller than a given threshold (and their corresponding eigenvectors)
+    are discarded.
+
+    The algorithm has the following steps [`arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`_]:
```

Comment:
Maybe the algorithm and theory could be combined together to make the overall docstring more intuitive.

Reply:
That will require extra explanation and makes the docstring very long, that is why I decided to have a compact algo at the end.